### PR TITLE
PLFM-6020 - Create GoogleCloudFileHandles using a file that has already been uploaded to Google Cloud.

### PR DESCRIPTION
**Pull request by nickgros**
https://sagebionetworks.jira.com/browse/PLFM-6020

### CODE REVIEW

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClient.java [hunk 1]
```diff
@@ -635,7 +636,15 @@ ExternalObjectStoreFileHandle createExternalObjectStoreFileHandle(ExternalObject
 	 * @throws SynapseException 
 	 */
 	public S3FileHandle createExternalS3FileHandle(S3FileHandle handle) throws SynapseException;
-	
+
+	/**
+	 * Create an S3FileHandle using a pre-configured ExternalS3StorageLocationSetting ID.
```
Comment (nickgros) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClient.java]: Fix doc

File: integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java [hunk 2]
```diff
@@ -558,6 +558,39 @@ public void testMultipartUploadV2Reset() throws FileNotFoundException, SynapseEx
...
+		fh.setStorageLocationId(storageLocationSetting.getStorageLocationId());
+		fh.setBucketName(config.getS3Bucket());
+		fh.setKey(key);
+		fh.setContentMd5(md5);
+
+		S3FileHandle result = synapse.createExternalS3FileHandle(fh);
+		assertNotNull(result);
+		toDelete.add(result);
+		assertNotNull(result.getFileName());
+	}
```
Comment (nickgros) [inline integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java]: Prior to this PR, there was no integration test for creating an S3FileHandle using an existing file in S3

File: integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java [hunk 3]
```diff
@@ -558,6 +558,39 @@ public void testMultipartUploadV2Reset() throws FileNotFoundException, SynapseEx
...
+		String md5 = MD5ChecksumHelper.getMD5Checksum(imageFile);
+
+		// Upload the owner.txt to S3 so we can create the external storage location
+		String baseKey = "integration-test/IT049FileHandleTest/testCreateExternalS3FileHandleFromExistingFile/" + UUID.randomUUID().toString();
+		String key = baseKey + FILE_NAME;
+		uploadOwnerTxtToS3(config.getS3Bucket(), baseKey, synapse.getUserProfile(userToDelete.toString()).getUserName());
+
+		// upload the little image to S3, but not through Synapse
+		synapseS3Client.putObject(config.getS3Bucket(), key, imageFile);
+
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java:576]: The object does not necessarily exist in S3 at the time the client returns from the putObject() call.  See other cases in which putObject() is called to see how we wait for the object to appear before continuing.

File: integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java [hunk 4]
```diff
@@ -558,6 +558,39 @@ public void testMultipartUploadV2Reset() throws FileNotFoundException, SynapseEx
...
+		storageLocationSetting.setUploadType(UploadType.S3);
+		storageLocationSetting = synapse.createStorageLocationSetting(storageLocationSetting);
+
+		S3FileHandle fh = new S3FileHandle();
+		fh.setStorageLocationId(storageLocationSetting.getStorageLocationId());
+		fh.setBucketName(config.getS3Bucket());
+		fh.setKey(key);
+		fh.setContentMd5(md5);
+
+		S3FileHandle result = synapse.createExternalS3FileHandle(fh);
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java:591]: add "// call under test"

File: integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java [hunk 5]
```diff
@@ -619,6 +652,42 @@ public void testMultipartUploadV2ToGoogleCloud() throws FileNotFoundException, S
...
+		String md5 = MD5ChecksumHelper.getMD5Checksum(imageFile);
+
+		// Upload the owner.txt to Google Cloud so we can create the external storage location
+		String baseKey = "integration-test/IT049FileHandleTest/testCreateExternalGoogleCloudFileHandleFromExistingFile/" + UUID.randomUUID().toString();
+		String key = baseKey + FILE_NAME;
+		uploadOwnerTxtToGoogleCloud(googleCloudBucket, baseKey, synapse.getUserProfile(userToDelete.toString()).getUserName());
+
+		// upload the little image to Google Cloud, but not through Synapse
+		googleCloudStorageClient.putObject(googleCloudBucket, key, imageFile);
+
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java:674]: wait for object to appear

File: integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java [hunk 6]
```diff
@@ -619,6 +652,42 @@ public void testMultipartUploadV2ToGoogleCloud() throws FileNotFoundException, S
...
+		storageLocationSetting.setUploadType(UploadType.GOOGLECLOUDSTORAGE);
+		storageLocationSetting = synapse.createStorageLocationSetting(storageLocationSetting);
+
+		GoogleCloudFileHandle fh = new GoogleCloudFileHandle();
+		fh.setStorageLocationId(storageLocationSetting.getStorageLocationId());
+		fh.setBucketName(googleCloudBucket);
+		fh.setKey(key);
+		fh.setContentMd5(md5);
+
+		GoogleCloudFileHandle result = synapse.createExternalGoogleCloudFileHandle(fh);
```
Comment (brucehoff) [inline integration-test/src/test/java/org/sagebionetworks/IT049FileHandleTest.java:689]: method under test

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java [hunk 7]
```diff
@@ -1096,6 +1097,63 @@ public S3FileHandle createExternalS3FileHandle(UserInfo userInfo,
...
+		 */
+		if(!esls.getCreatedBy().equals(userInfo.getId())){
+			throw new UnauthorizedException("Only the creator of ExternalGoogleCloudStorageLocationSetting.id="+fileHandle.getStorageLocationId()+" can create an external GoogleCloudFileHandle with storageLocationId = "+fileHandle.getStorageLocationId());
+		}
+		try {
+			Blob summary = googleCloudStorageClient.getObject(fileHandle.getBucketName(), fileHandle.getKey());
+			if (fileHandle.getContentSize() == null) {
+				fileHandle.setContentSize(summary.getSize());
+			}
+		} catch (Exception e) {
```
Comment (brucehoff) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java:1144]: either (1) replace 'Exception' with the particular exception type thrown by 'getObject' or (2) ensure that only the getObject() method invocation is within the try/catch block.

File: services/repository/src/main/java/org/sagebionetworks/file/controller/UploadController.java [hunk 8]
```diff
@@ -360,6 +361,35 @@
...
+	 *
+	 * @param userId
+	 * @param fileHandle
+	 *            The GoogleCloudFileHandle to create
+	 * @return
+	 * @throws DatastoreException
+	 * @throws NotFoundException
+	 */
+	@ResponseStatus(HttpStatus.OK)
+	@RequestMapping(value = "/externalFileHandle/googleCloud", method = RequestMethod.POST)
```
Comment (brucehoff) [inline services/repository/src/main/java/org/sagebionetworks/file/controller/UploadController.java]: the constant "/externalFileHandle/googleCloud" should be defined in a file (URLHelpers.java??)

**Review by brucehoff — APPROVED** [review]
Looks great!

**Pull request merged**